File: sorting/quick_sort/quick_sort.js
// Quick sort algorithm
// Quicksort has a time complexity of O(nlog n)

// quickSort([11,8,14,3,6,2,7],0,6); 
// //[2, 3, 6, 7, 8, 11, 14]
// quickSort([11,8,14,3,6,2,1, 7],0,7);
// //[1, 2, 3, 6, 7, 8, 11, 14]
// quickSort([16,11,9,7,6,5,3, 2],0,7);
// //[2, 3, 5, 6, 7, 9, 11, 16]


function partition(arr, pivot, left, right){
   var pivotValue = arr[pivot],
       partitionIndex = left;

   for(var i = left; i < right; i++){
    if(arr[i] < pivotValue){
      swap(arr, i, partitionIndex);
      partitionIndex++;
    }
  }
  swap(arr, right, partitionIndex);
  return partitionIndex;
}

function swap(arr, i, j){
   var temp = arr[i];
   arr[i] = arr[j];
   arr[j] = temp;
}
    
function quickSort(arr, left, right){
   var len = arr.length, 
   pivot,
   partitionIndex;


  if(left < right){
    pivot = right;
    partitionIndex = partition(arr, pivot, left, right);
    
   //sort left and right
   quickSort(arr, left, partitionIndex - 1);
   quickSort(arr, partitionIndex + 1, right);
  }
  return arr;
}

File: sorting/quick_sort/quick_sort.py
import random
import time
def partition(arr,low,high):
    i = ( low-1 )         
    pivot = arr[high]     
    for j in range(low , high):       
        if   arr[j] <= pivot:           
            i = i+1 
            arr[i],arr[j] = arr[j],arr[i]
    arr[i+1],arr[high] = arr[high],arr[i+1]
    return ( i+1 )
def quickSort(arr,low,high):
    if low < high:        
        pi = partition(arr,low,high)        
        quickSort(arr, low, pi-1)
        quickSort(arr, pi+1, high)
arr = random.sample(range(10000),10)
n = len(arr)
x = time.time()
quickSort(arr,0,n-1)
y = time.time()
print("Time taken by quick sort is: ")
print(y-x)
print ("Sorted array is:")
for i in range(n):
    print ("%d" %arr[i])
